File: python/dnd-character/dnd_character.py
"""exercise"""
import random
import math

class Character:
    """
    present a rolled character sheet
    """
    def __init__(self):
        self.strength = self.ability()
        self.dexterity = self.ability()
        self.constitution = self.ability()
        self.intelligence = self.ability()
        self.wisdom= self.ability()
        self.charisma= self.ability()
        self.hitpoints = modifier(self.constitution) + 10

    def ability(self):
        """
        :return: a valid ability score
        """
        rolls = [random.randrange(1,6) for _ in range(0,4)]
        rolls.sort()
        return sum(rolls[1::])


def modifier(stat):
    """
    :param stat:
    :return:
    """
    return math.floor((stat - 10) / 2)

File: lua/dnd-character/dnd-character.lua
local Character = {}

local function roll_dice()
  local rolls = {}
  for i = 1,4  do
    rolls[i]  = math.random(1,6)
  end
  return rolls
end

local function ability(scores)
  if scores == nil then
    scores = roll_dice()
  end

  table.sort(scores)

  local sum = 0
  for i = 2, 4 do
    sum = sum + scores[i]
  end

  return sum
end

local function modifier(input)
  local mod = 0
  if input > 17 then
    mod = 4
  elseif input > 15 then
    mod = 3
  elseif input > 13 then
    mod = 2
  elseif input > 11 then
    mod = 1
  elseif input > 9 then
    mod = 0
  elseif input > 7 then
    mod = -1
  elseif input > 5 then
    mod = -2
  elseif input > 3 then
    mod = -3
  else
    mod = -4
  end

  return mod
end

function Character:new(name)
  local con = ability(nil)
  local c = {
    name = name,
    strength = ability(nil),
    dexterity = ability(nil),
    constitution = con,
    intelligence = ability(nil),
    wisdom = ability(nil),
    charisma = ability(nil),
    hitpoints = 10 + modifier(con)
  }
  return c
end

return {
  Character = Character,
  ability = ability,
  roll_dice = roll_dice,
  modifier = modifier
}
